### ISSUE TITLE
Problem with maven Surefire plugin and Java 11

### ISSUE THREAD

**Issue by johann-petrak** [GateNLP/gateplugin-Format_Bdoc#8]
Testing aborts when running under Java 11. 

Java:
```
openjdk version "11.0.8" 2020-07-14
OpenJDK Runtime Environment 18.9 (build 11.0.8+10)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.8+10, mixed mode)
```

This refers to commit f12eda938979ca43cc02b74c2e5f7c8c9b4eed78

Console output:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.0:test (unit-tests) on project format-bdoc: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/johann/gate-git/gateplugin-Format_Bdoc/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date]-jvmRun[N].dump, [date].dumpstream and [date]-jvmRun[N].dumpstream.
[ERROR] The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
[ERROR] Command was /bin/sh -c cd /home/johann/gate-git/gateplugin-Format_Bdoc && /home/johann/.sdkman/candidates/java/11.0.8-open/bin/java -javaagent:/home/johann/.m2/repository/org/jacoco/org.jacoco.agent/0.8.1/org.jacoco.agent-0.8.1-runtime.jar=destfile=/home/johann/gate-git/gateplugin-Format_Bdoc/target/jacoco.exec -jar /home/johann/gate-git/gateplugin-Format_Bdoc/target/surefire/surefirebooter13816448777889195805.jar /home/johann/gate-git/gateplugin-Format_Bdoc/target/surefire 2020-09-07T10-07-36_913-jvmRun1 surefire12237597699839995081tmp surefire_09045107206883320244tmp
[ERROR] Error occurred in starting fork, check output in log
[ERROR] Process Exit Code: 134
[ERROR] org.apache.maven.surefire.booter.SurefireBooterForkException: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
[ERROR] Command was /bin/sh -c cd /home/johann/gate-git/gateplugin-Format_Bdoc && /home/johann/.sdkman/candidates/java/11.0.8-open/bin/java -javaagent:/home/johann/.m2/repository/org/jacoco/org.jacoco.agent/0.8.1/org.jacoco.agent-0.8.1-runtime.jar=destfile=/home/johann/gate-git/gateplugin-Format_Bdoc/target/jacoco.exec -jar /home/johann/gate-git/gateplugin-Format_Bdoc/target/surefire/surefirebooter13816448777889195805.jar /home/johann/gate-git/gateplugin-Format_Bdoc/target/surefire 2020-09-07T10-07-36_913-jvmRun1 surefire12237597699839995081tmp surefire_09045107206883320244tmp
[ERROR] Error occurred in starting fork, check output in log
[ERROR] Process Exit Code: 134
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:671)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:533)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:278)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:244)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:1194)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:1022)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:868)
[ERROR] 	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
[ERROR] 	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
[ERROR] 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
[ERROR] 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
[ERROR] 	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
[ERROR] 	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:957)
[ERROR] 	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:289)
[ERROR] 	at org.apache.maven.cli.MavenCli.main(MavenCli.java:193)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
[ERROR] 
```

This is the content of the surefire dumpstream file:
```
# Created at 2020-09-07T10:07:37.360
Exception in thread "main" java.lang.reflect.InvocationTargetException

# Created at 2020-09-07T10:07:37.361
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)

# Created at 2020-09-07T10:07:37.361
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)

# Created at 2020-09-07T10:07:37.361
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

# Created at 2020-09-07T10:07:37.362
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)

# Created at 2020-09-07T10:07:37.362
	at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:513)

# Created at 2020-09-07T10:07:37.362
	at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:525)

# Created at 2020-09-07T10:07:37.362
Caused by: java.lang.RuntimeException: Class java/lang/UnknownError could not be instrumented.

# Created at 2020-09-07T10:07:37.362
	at org.jacoco.agent.rt.internal_c13123e.core.runtime.ModifiedSystemClassRuntime.createFor(ModifiedSystemClassRuntime.java:140)

# Created at 2020-09-07T10:07:37.362
	at org.jacoco.agent.rt.internal_c13123e.core.runtime.ModifiedSystemClassRuntime.createFor(ModifiedSystemClassRuntime.java:101)

# Created at 2020-09-07T10:07:37.363
	at org.jacoco.agent.rt.internal_c13123e.PreMain.createRuntime(PreMain.java:55)

# Created at 2020-09-07T10:07:37.363
	at org.jacoco.agent.rt.internal_c13123e.PreMain.premain(PreMain.java:47)

# Created at 2020-09-07T10:07:37.363
	... 6 more

# Created at 2020-09-07T10:07:37.363
Caused by: java.lang.NoSuchFieldException: $jacocoAccess

# Created at 2020-09-07T10:07:37.363
	at java.base/java.lang.Class.getField(Class.java:1999)

# Created at 2020-09-07T10:07:37.363
	at org.jacoco.agent.rt.internal_c13123e.core.runtime.ModifiedSystemClassRuntime.createFor(ModifiedSystemClassRuntime.java:138)

# Created at 2020-09-07T10:07:37.363
	... 9 more

# Created at 2020-09-07T10:07:37.363
*** java.lang.instrument ASSERTION FAILED ***: "result" with message agent load/premain call failed at ./src/java.instrument/share/native/libinstrument/JPLISAgent.c line: 422

# Created at 2020-09-07T10:07:37.488
Aborted (core dumped)
```

**Comment by johann-petrak** [comment]
Related SO https://stackoverflow.com/questions/55632614/maven-surefire-plugin-crashing-jvm-on-java-11-corrupted-stdout-by-directly-writ/60795159

**Comment by johann-petrak** [comment]
The recommended solution using forkNode is committed as adba27b0b5ac1b2f2221bf81ada5d92700036604 

When doing `mvn clean compile test`  it hangs forever during the testing stage

**Comment by johann-petrak** [comment]
Non-recommended solution using forkCount=0 is committed as dd8793f08b114d3841c28ec9271222bba453ea44 and works

**Comment by johann-petrak** [comment]
To reproduce, since this depends on SNAPSHOT versions, the following maven config file has to be added as `~/.m2/settings.xml`

```
<settings>
  <profiles>
    <profile>
      <id>gate-development</id>
      <repositories>
        <repository>
          <id>gate-maven-snapshots</id>
          <name>GATE Snapshot Repository</name>
          <url>http://repo.gate.ac.uk/content/repositories/snapshots/</url>
        </repository>
      </repositories>
    </profile>
  </profiles>
  <activeProfiles>
    <activeProfile>gate-development</activeProfile>
  </activeProfiles>
</settings>
```

**Comment by Tibor17** [comment]
@johann-petrak Of course that you don't have a problem with forkCount=0 because the JaCoCo cannot be used there. But I can say that this is an issue with JaCoCo. It is old issue in JaCoCo which I saw in my project `Caused by: java.lang.NoSuchFieldException: $jacocoAccess` but it has nothing to do with Surefire. I could not see how you use JaCoCo in your POM. Can you show me?

**Comment by johann-petrak** [comment]
Thank you @Tibor17 for getting back here! TBH I do not know much about the details either since I am using a parent project pom created by other people. The parent pom is here: https://github.com/GateNLP/gate-top/blob/d4e14a11546b201b9d11e1dda28d5a1e7271b9cc/gate-plugin-base/pom.xml#L179

TBH I do not know what really goes on here: if there is a problem with JaCoCo why does it only show unless I set forkCount to 0?

**Comment by Tibor17** [comment]
@johann-petrak 
I want to know how you added this
`-javaagent:/home/johann/.m2/repository/org/jacoco/org.jacoco.agent/0.8.1/org.jacoco.agent-0.8.1-runtime.jar`
In the Jenkins?
It is normally done in POM but you don't have this approach. I bet the version 0.8.1 is buggy. Pls use 0.8.5. It is cool stable for me and I am using all since Java 1.7 through Java 11 - Java 14 and Java 15 without any problems.
If it helps can you please rank my [stackoverflow answer](https://stackoverflow.com/questions/55632614) back to high if my answer helps? Thx

**Comment by Tibor17** [comment]
Is Jenkins overriding the version from 0.8.5 to 0.8.1?

**Comment by johann-petrak** [comment]
I am running this from my terminal on my laptop, no Jenkins involved. So I am totally confused why that Jacoco version is showing up at all. 

Who is actually creating the comman line for that subprocess?

**Comment by ianroberts** [comment]
@johann-petrak the master branch of format-bdoc points to `gate-plugin-base` 8.6.1, which depends on jacoco 0.8.1, not the 9.0-SNAPSHOT one you linked to above.

**Comment by johann-petrak** [comment]
OK, so I cannot depend on a non-SNAPSHOT version of the gate-plugin-base and compile with Java 11?

**Comment by johann-petrak** [comment]
I still do not understand why it looks like I can compile OK with forkCount=0 and why my build hangs if I use `<forkNode implementation="org.apache.maven.plugin.surefire.extensions.SurefireForkNodeFactory"/>` instead though.

**Comment by greenwoodma** [comment]
You can build you just can't run the tests. We should probably fix this sooner rather than later, but given we still use Java 8 for building releases (and recommend that for others building plugins etc. to ensure compatibility) it's not a huge priority; over other work rather than not a priority at all.

**Comment by greenwoodma** [comment]
> I still do not understand why it looks like I can compile OK with forkCount=0 and why my build hangs if I use `<forkNode implementation="org.apache.maven.plugin.surefire.extensions.SurefireForkNodeFactory"/>` instead though.

That's because when you set `forkCount=0` it doesn't fork a new process to run the tests so it can't add the jacoco user agent to the jvm so it doesn't fail.

**Comment by Tibor17** [comment]
In my long term experience with JaCoCo, the bug `NoSuchFieldException: $jacocoAccess` was seen in old version of JaCoCo and Java 11 or Java 13.
Pls upgrade the version of JaCoCo and let me know the build result. Thx

**Comment by Tibor17** [comment]
> 
> 
> I still do not understand why it looks like I can compile OK with forkCount=0 and why my build hangs if I use `<forkNode implementation="org.apache.maven.plugin.surefire.extensions.SurefireForkNodeFactory"/>` instead though.

The answer is simple, so the runtime of forked JVM is broken by jacoco library and the plugins does not see the JVM exit. Try to log the stack trace and we will see more. But still we can se that the jacoco version must be updated.

Your Jenkins does not crash because this bug appears with Java 11+ only.
Your build system uses Java 1.8 and therefore this precondition is fine for 0.8.1.
If you turn the build system to Java 11, you will see it crashed in Jenkins too.

**Comment by Tibor17** [comment]
@ianroberts 
@johann-petrak 
Have you updated the parent POM, i.e. `gate-plugin-base`, with the latest JaCoCo version 0.8.5?
How is it?
I want to make sure that the problem is only in the jacoco version and not in the surefire.

**Comment by johann-petrak** [comment]
@Tibor17 there is an updated (SNAPSHOT) version and I tested it with that one and it works. Thank you for all your help!

**Comment by johann-petrak** [comment]
Conclusion: document that for running tests we need to stick with Java 8 for now as long as we want to depend on the released GATE 8.6.1 version.
